session/nontransactional.go

```diff
@@ -621,6 +704,46 @@ func selectShardColumn(stmt *ast.NonTransactionalDMLStmt, se Session, tableName
 	return indexed, shardColumnInfo, nil
 }
 
+func selectShardColumnAutomatically(stmt *ast.NonTransactionalDMLStmt, tbl table.Table,
+	tableName *ast.TableName, tableAsName model.CIStr) (bool, *model.ColumnInfo, error) {
+	// auto-detect shard column
+	var shardColumnInfo *model.ColumnInfo
+	tableInfo := tbl.Meta()
+	if tbl.Meta().PKIsHandle {
+		shardColumnInfo = tableInfo.GetPkColInfo()
+	} else if tableInfo.IsCommonHandle {
+		for _, index := range tableInfo.Indices {
+			if index.Primary {
+				if len(index.Columns) == 1 {
+					shardColumnInfo = tableInfo.Columns[index.Columns[0].Offset]
+					break
+				}
+				// if the clustered index contains multiple columns, we cannot automatically choose a column as the shard column
+				return false, nil, errors.New("Non-transactional DML, the clustered index contains multiple columns. Please specify a shard column")
```

Comment:
Should this be a TODO? Multi-column clustered index seems doable.

Reply:
Do you mean by selecting the first column? That column might be highly skewed. So I'd rather let user manually specify it.

> Is there any test for selecting sharding key automatically with clustered index?

It's in `TestNonTransactionalWithAutoDetectShardColumn`

Reply:
Single column can be highly skewed too. I don't get why multi-column is special.

I thought you didn't do it just because of implementation...

Reply:
Ah, actually I don't mean using first column only. To support multiple columns, we need to scan batches of keys containing all columns in the clustered index.